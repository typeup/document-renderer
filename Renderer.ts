import * as dom from "@typeup/dom"
export type renderFunction = (renderer: Renderer, node: dom.Node) => Promise<string>

export class Renderer {
	private constructor(private variables: { [name: string]: string } = {}) {
	}
	getVariable(name: string): string {
		return this.variables[name]
	}
	setVariable(name: string, value: string) {
		this.variables[name] = value
	}
	async render(node: dom.Node | dom.Node[]): Promise<string> {
		let result: string
		if (Array.isArray(node))
			result = (await Promise.all(node.map(n => this.render(n)))).reduce<string[]>((r, e) => Array.isArray(e) ? [...r, ...e] : [...r, e], []).join("")
		else {
			const render = renderers[node.class]
			result = render ? await render(this, node) : ""
		}
		return result
	}
	scope(): Renderer {
		return new Renderer({ ...this.variables })
	}
	static create(): Renderer {
		return new Renderer({})
	}
}
const renderers: { [className: string]: renderFunction } = {}
export function addRenderer(className: string, render: renderFunction) {
	renderers[className] = render
}
