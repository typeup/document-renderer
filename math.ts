// tslint:disable: no-submodule-imports
import { mathjax } from "mathjax-full/js/mathjax"
import { TeX } from "mathjax-full/js/input/tex"
import { CHTML } from "mathjax-full/js/output/chtml"
import { liteAdaptor } from "mathjax-full/js/adaptors/liteAdaptor"
import { RegisterHTMLHandler } from "mathjax-full/js/handlers/html"
import { AllPackages } from "mathjax-full/js/input/tex/AllPackages"

export async function typeset(document: string): Promise<string> {
	const adaptor = liteAdaptor()
	RegisterHTMLHandler(adaptor as any)
	const tex = new TeX({ packages: AllPackages, inlineMath: [["$", "$"]] })
	const chtml = new CHTML({ fontURL: "https://cdnjs.cloudflare.com/ajax/libs/mathjax/3.0.0/es5/output/chtml/fonts/woff-v2" })
	const html = mathjax.document(document, { InputJax: tex, OutputJax: chtml, })
	html.render()
	return adaptor.outerHTML(adaptor.root(html.document))
}
