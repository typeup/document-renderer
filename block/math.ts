import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"
import * as MathJaxNode from "mathjax-node"

async function render(renderer: Renderer, me: dom.block.Math): Promise<string> {
	MathJaxNode.config({
		MathJax: {
		},
	})
	MathJaxNode.start()
	const result = await new Promise<string>((resolve, reject) => MathJaxNode.typeset({
		math: me.value,
		format: "inline-TeX",
		mml: true,
	}, data => {
		if (data.errors)
			reject(data.errors)
		else
			resolve(data.mml)
	}))
	return `<figure>
	${result}
	<figcaption>${ await renderer.render(me.content) }</figcaption>
</figure>
`
}
addRenderer("Block.Math", render)
