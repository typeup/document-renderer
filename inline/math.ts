import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"

async function render(renderer: Renderer, me: dom.inline.Math): Promise<string> {
	return `$${ me.value }$`
}
addRenderer("Inline.Math", render)
