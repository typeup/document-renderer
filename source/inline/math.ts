import * as dom from "@typeup/dom"
import { Renderer, addRenderer } from "../Renderer"
import * as MathJaxNode from "mathjax-node"

async function render(renderer: Renderer, me: dom.inline.Link): Promise<string> {
	MathJaxNode.config({
		MathJax: {
		},
	})
	MathJaxNode.start()
	const result = await new Promise<string>((resolve, reject) => MathJaxNode.typeset({
		math: me.content,
		format: "inline-TeX",
		mml: true,
	}, data => {
		if (data.errors)
			reject(data.errors)
		else
			resolve(data.mml)
	}))
	return result
}
addRenderer("Inline.Math", render)
