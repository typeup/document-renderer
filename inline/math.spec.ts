import * as dom from "@typeup/dom"
import { Renderer } from "../Renderer"
import "./math"

const renderer = Renderer.create()
describe("Inline.Math", () => {
	it("render", async () => {
		expect(await renderer.render(new dom.inline.Math("a^2 + b^2 = c^2"))).toMatchSnapshot()
	})
})
