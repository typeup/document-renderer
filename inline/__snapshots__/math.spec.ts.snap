// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline.Math render 1`] = `"$a^2 + b^2 = c^2$"`;
